internal/functions/efgs/upload-batch.go

```diff
@@ -104,47 +104,32 @@ func UploadBatch(w http.ResponseWriter, r *http.Request) {
 
 func uploadBatchConfiguration(ctx context.Context) (*uploadConfiguration, error) {
 	logger := logging.FromContext(ctx).Named("efgs.uploadBatchConfiguration")
-	secretsClient := secrets.Client{}
 
-	var err error
-	var config uploadConfiguration
+	efgsEnv := efgsutils.GetEfgsEnvironmentOrFail()
 
-	config.NBTLSPair, err = efgsutils.LoadX509KeyPair(ctx, efgsutils.NBTLS)
-	if err != nil {
-		logger.Debug("Error loading authentication certificate")
-		return nil, err
-	}
+	url := efgsutils.GetEfgsURLOrFail(efgsEnv)
```

Comment:
can we please use Dependency Injection everywhere? We use it in like half of the cases and the codebase is very inconsistent. This approach of mutating the internal state inside of the functions based on global vars or calling other functions is impossible to unit test. 

I'd recommend reading like the beginning of this blogpost for example https://medium.com/@elliotchance/a-new-simpler-way-to-do-dependency-injection-in-go-9e191bef50d5

Reply:
This is impossible to unit test anyway. There are so many things missing for that goal that I don't believe this one makes any difference...  
There's a function named `uploadBatch` below and it receives the configuration (that is created here on **this** line)
```go
type uploadConfiguration struct {
	URL       *urlutils.URL
	Env       efgsutils.Environment
	NBTLSPair *efgsutils.X509KeyPair
	Client    *http.Client
	BatchTag  string
}

func uploadBatch(ctx context.Context, batch *efgsapi.DiagnosisKeyBatch, config *uploadConfiguration) (*efgsapi.UploadBatchResponse, error) { ... }
```
Even `Client` is there. The only thing that is untestable here is database handling - ok, I can put even the `Database` instance inside of the `uploadConfiguration`.

Reply:
I know how to create testable code. It's just very hard to do such thing in a language like Go. The linked article didn't convince me about anything but the Go is very screwed language for writing complex applications.

Reply:
Our application is far from complex. We use Go and there is nothing we can do about it anyway. What we can do however is use DI and write testable functions.

Reply:
Yes, it's very far, yet complex enough :smile:

Reply:
The whole EFGS code was refactored to be more testable.